File: generator/src/pipeline/consensus.ts
/**
 * BetaTetris normal-net top-1 **consensus filter** (#55) — the standard *final*
 * stage of bank generation (docs/decisions.md, 2026-06-21).
 *
 * Generation stays pure TS/StackRabbit (see {@link generateBank}); this stage
 * runs afterwards and drops any puzzle the BetaTetris **normal** net does not
 * bless: a puzzle is kept iff the net's #1 piece-1 policy move lands the SAME
 * board as our stored optimal's piece-1 placement (compared by the canonical
 * 200-char outcome key, so the check is convention-free). It is a **filter, not
 * a re-rank** — disagreers are dropped, never relabelled with BT's move, so the
 * StackRabbit-derived combo table and graded rewards stay the optimal's.
 *
 * Two guardrails are baked in:
 *  - **Normal net only.** The `perfect` net is off-objective for a general
 *    stacking trainer (trained for maxout/killscreen tetris-only play) and is
 *    dropped from the standard path.
 *  - **Fail-closed.** A puzzle BetaTetris cannot cleanly judge (engine error,
 *    unreachable outcome, odd-parity board, inject mismatch) is rejected, and the
 *    BT-error count is surfaced separately so flakiness never silently inflates
 *    the cull.
 *
 * BetaTetris is offline / generator-only (the StackRabbit guardrail applies
 * equally — never deployed, never called from `apps/play`; GPLv3). The actual
 * net is reached through an injectable {@link ConsensusJudge}, so this stage is
 * unit-tested without PyTorch; the production judge ({@link betaTetrisJudge})
 * shells to `betatetris-spike/consensus.py`.
 */

import { spawn } from 'node:child_process';
import { mkdtempSync, readFileSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { applyPlacement, boardKey, decodeBoard, type Line } from '@trainer/core';

/** The minimum a puzzle must expose to be consensus-checked. */
export interface ConsensusPuzzle {
  /** Stable id (DB uuid for live puzzles; any tag for fresh ones). */
  id?: string;
  number?: number | null;
  /** 200-char encoded start board (`encodeBoard`). */
  board: string;
  piece1: string;
  piece2: string;
  /** The stored optimal two-placement line (rank-1 combo). */
  optimalLine: Line;
}

/** Outcome keys after the stored optimal's piece-1 (and both) placements. */
export function consensusKeys(p: ConsensusPuzzle): { p1Key: string; fullKey: string } {
  const board0 = decodeBoard(p.board);
  const afterP1 = applyPlacement(board0, p.piece1 as never, p.optimalLine[0]);
  const afterP2 = applyPlacement(afterP1, p.piece2 as never, p.optimalLine[1]);
  return { p1Key: boardKey(afterP1), fullKey: boardKey(afterP2) };
}

/** One keyed puzzle as handed to the BT judge (the `bank_keys.json` row shape). */
export interface ConsensusKeyRow {
  id: string;
  number: number | null;
  board: string;
  piece1: string;
  piece2: string;
  /** Outcome key after the optimal's piece-1 placement — the move BT must bless. */
  p1_key: string;
  full_key: string;
}

/** Why a puzzle was dropped (or `null` when kept). `bt-error` is machinery failure. */
export type ConsensusReason =
  | 'disagree'
  | 'unreachable'
  | 'odd-parity'
  | 'inject-mismatch'
  | 'bt-error';

/** The BT verdict for one puzzle. */
export interface ConsensusVerdict {
  number: number | null;
  id: string | null;
  keep: boolean;
  reason: ConsensusReason | null;
  /** 1-indexed rank of our optimal in the net's policy, or `null` if unreachable. */
  rank: number | null;
}

/**
 * The BetaTetris surface: judge a batch of keyed puzzles, returning one verdict
 * per row in the SAME order. Injected so the filter is testable without the net.
 */
export type ConsensusJudge = (rows: ConsensusKeyRow[]) => Promise<ConsensusVerdict[]>;

/** Outcome of a consensus pass over a set of puzzles. */
export interface ConsensusResult<T extends ConsensusPuzzle> {
  /** Puzzles the net blessed (top-1 consensus). */
  kept: T[];
  /** Dropped puzzles with the reason each failed. */
  dropped: Array<{ puzzle: T; reason: ConsensusReason }>;
  /** Fraction kept (genuine top-1 consensus rate, BT-errors included in the denominator). */
  keepRate: number;
  /** Drops attributable to machinery failure (logged apart from genuine disagree). */
  btErrors: number;
  /** The raw per-puzzle verdicts, in input order. */
  verdicts: ConsensusVerdict[];
}

/**
 * Partition `puzzles` into kept / dropped by the BetaTetris top-1 consensus
 * verdict (#55). Fail-closed: a missing or malformed verdict is treated as a
 * `bt-error` drop, so a flaky judge can only ever shrink the bank, never keep an
 * unjudged puzzle. The kept set is guaranteed 100% top-1-consensus.
 */
export async function filterByConsensus<T extends ConsensusPuzzle>(
  puzzles: T[],
  judge: ConsensusJudge,
): Promise<ConsensusResult<T>> {
  if (puzzles.length === 0) {
    return { kept: [], dropped: [], keepRate: 1, btErrors: 0, verdicts: [] };
  }
  const rows: ConsensusKeyRow[] = puzzles.map((p, i) => {
    const { p1Key, fullKey } = consensusKeys(p);
    return {
      id: p.id ?? String(i),
      number: p.number ?? null,
      board: p.board,
      piece1: p.piece1,
      piece2: p.piece2,
      p1_key: p1Key,
      full_key: fullKey,
    };
  });

  const verdicts = await judge(rows);

  const kept: T[] = [];
  const dropped: Array<{ puzzle: T; reason: ConsensusReason }> = [];
  let btErrors = 0;
  const out: ConsensusVerdict[] = [];
  for (let i = 0; i < puzzles.length; i++) {
    const v = verdicts[i];
    // Fail-closed: no verdict (judge returned short / crashed) → bt-error drop.
    if (!v) {
      dropped.push({ puzzle: puzzles[i], reason: 'bt-error' });
      btErrors++;
      out.push({ number: rows[i].number, id: rows[i].id, keep: false, reason: 'bt-error', rank: null });
      continue;
    }
    out.push(v);
    if (v.keep) {
      kept.push(puzzles[i]);
    } else {
      const reason = v.reason ?? 'bt-error';
      if (reason === 'bt-error') btErrors++;
      dropped.push({ puzzle: puzzles[i], reason });
    }
  }

  return {
    kept,
    dropped,
    keepRate: kept.length / puzzles.length,
    btErrors,
    verdicts: out,
  };
}

/** Where the production Python adapter lives, relative to the repo root. */
const CONSENSUS_PY = 'betatetris-spike/consensus.py';

/**
 * The production {@link ConsensusJudge}: write the keyed rows to a temp file,
 * run `bt-run python consensus.py` (the normal-net top-1 verdict) over them, and
 * read back the per-puzzle verdicts (matched to input rows by id). Offline /
 * generator-only. Pass `{ cwd }` for the repo root if not the process cwd.
 */
export function betaTetrisJudge(opts: { cwd?: string; runner?: string } = {}): ConsensusJudge {
  const cwd = opts.cwd ?? process.cwd();
  const runner = opts.runner ?? 'bt-run';
  return async (rows: ConsensusKeyRow[]): Promise<ConsensusVerdict[]> => {
    const dir = mkdtempSync(join(tmpdir(), 'bt-consensus-'));
    const inPath = join(dir, 'keys.json');
    const outPath = join(dir, 'verdict.json');
    writeFileSync(inPath, JSON.stringify(rows));

    await new Promise<void>((resolve, reject) => {
      const child = spawn(runner, ['python', CONSENSUS_PY, inPath, outPath], {
        cwd,
        stdio: ['ignore', 'inherit', 'inherit'],
      });
      child.on('error', reject);
      child.on('close', (code) =>
        code === 0 ? resolve() : reject(new Error(`consensus.py exited ${code}`)),
      );
    });

    const raw = JSON.parse(readFileSync(outPath, 'utf8')) as ConsensusVerdict[];
    // Re-key by id so order/length mismatches fail closed (missing → undefined →
    // the filter treats it as a bt-error drop).
    const byId = new Map(raw.map((v) => [v.id, v]));
    return rows.map((r) => byId.get(r.id) as ConsensusVerdict);
  };
}

File: engines/betatetris/consensus.py
"""#55 — BetaTetris normal-net **top-1 consensus** verdict (the standard gate).

The production consensus filter (`generator/src/pipeline/consensus.ts`) shells to
this. For each puzzle it asks the BetaTetris **normal** net (v1.0.0):

 1. **Piece 1**: is our stored optimal's piece-1 outcome the net's #1 policy
    move? (same as before)
 2. **Piece 2**: given BT's own top-1 piece-1 placement, does BT's top-1
    piece-2 placement match our stored optimal's full outcome — for ALL 7
    possible next-pieces (p3)?  Our optimal is p3-agnostic; BT's policy sees
    p3 in the observation, so we sweep all 7 and require unanimous agreement.

Keep iff BOTH pass.

Reuses the measurement machinery from `keeprate.py` (board injection, the
two-phase adjustment cadence, the convention-free 200-char outcome keys) but
emits a per-puzzle keep/drop **verdict** instead of an aggregate keep-rate, and
runs the **normal net only** — the `perfect` net is off-objective for a general
stacking trainer (it is trained for maxout/killscreen tetris-only play) and is
dropped from the standard path (.claude/docs/decisions.md, 2026-06-21).

**Fail-closed:** a puzzle BetaTetris cannot cleanly judge is DROPPED, and the
reason distinguishes a genuine *disagree* (rank > 1) from machinery failure so a
flaky BT run never silently inflates the cull:
  - `disagree`        — piece-1 reachable but not the net's top-1.
  - `disagree-p2`     — piece-1 agrees but piece-2 disagrees for >=1 of 7 p3s.
  - `unreachable`     — our optimal outcome is not even in the net's move set.
  - `odd-parity`      — board parity BetaTetris's Reset cannot accept.
  - `inject-mismatch` — injected board0 round-trips wrong (engine/convention bug).
  - `bt-error`        — any exception while judging this puzzle.

  bt-run python engines/betatetris/consensus.py [keys.json] [out.json] [limit]

Defaults: keys = $BT_OUT/bank_keys.json, out = $BT_OUT/consensus_verdict.json.
Offline / generator-only (GPLv3 BetaTetris).
"""
import sys, os, json, time
import keeprate as kr  # reuse the measured, correct-cadence harness

P2_THRESHOLD = 7  # require all 7 p3 values to agree (strictest)


def verdict_for(model, pz):
    """keep/drop verdict for one puzzle under the normal net. Fail-closed."""
    number, pid = pz.get('number'), pz.get('id')
    lines = kr.lines_for(pz['board'])
    if lines is None:
        return {'number': number, 'id': pid, 'keep': False,
                'reason': 'odd-parity', 'rank': None, 'inject_ok': None,
                'p2_agree': None, 'p2_of': None}
    try:
        dist, inject_ok = kr.piece1_outcome_distribution(
            model, pz['board'], pz['piece1'], pz['piece2'], lines)
    except Exception as e:  # fail-closed: never let a BT crash keep a puzzle
        return {'number': number, 'id': pid, 'keep': False,
                'reason': 'bt-error', 'rank': None, 'inject_ok': None,
                'p2_agree': None, 'p2_of': None, 'error': str(e)[:200]}
    if not inject_ok:
        return {'number': number, 'id': pid, 'keep': False,
                'reason': 'inject-mismatch', 'rank': None, 'inject_ok': False,
                'p2_agree': None, 'p2_of': None}
    rank = kr.rank_of(dist, pz['p1_key'])
    if rank is None:
        reason = 'unreachable'
    elif rank == 1:
        reason = None
    else:
        reason = 'disagree'
    if reason is not None:
        return {'number': number, 'id': pid, 'keep': False, 'reason': reason,
                'rank': rank, 'inject_ok': True, 'n_outcomes': len(dist),
                'p2_agree': None, 'p2_of': None}

    # Piece 1 passed — check piece 2 across all 7 possible p3 values.
    try:
        p1_premove, p1_action, _ = kr.piece1_top1_actions(
            model, pz['board'], pz['piece1'], pz['piece2'], lines)
        p2_agree = 0
        for p3_id in range(7):
            p2_dist, topped = kr.piece2_outcome_distribution(
                model, pz['board'], pz['piece1'], pz['piece2'], p3_id, lines,
                p1_premove, p1_action)
            if topped:
                continue
            p2_rank = kr.rank_of(p2_dist, pz['full_key'])
            if p2_rank == 1:
                p2_agree += 1
    except Exception as e:
        return {'number': number, 'id': pid, 'keep': False,
                'reason': 'bt-error', 'rank': rank, 'inject_ok': True,
                'n_outcomes': len(dist), 'p2_agree': None, 'p2_of': 7,
                'error': str(e)[:200]}

    keep = p2_agree >= P2_THRESHOLD
    p2_reason = None if keep else 'disagree-p2'
    return {'number': number, 'id': pid, 'keep': keep, 'reason': p2_reason,
            'rank': rank, 'inject_ok': True, 'n_outcomes': len(dist),
            'p2_agree': p2_agree, 'p2_of': 7}


def main():
    out_dir = kr.BT_OUT
    keys_path = sys.argv[1] if len(sys.argv) > 1 else os.path.join(out_dir, 'bank_keys.json')
    out_path = sys.argv[2] if len(sys.argv) > 2 else os.path.join(out_dir, 'consensus_verdict.json')
    limit = int(sys.argv[3]) if len(sys.argv) > 3 else None

    bank = json.load(open(keys_path))
    if limit:
        bank = bank[:limit]
    print(f"consensus: judging {len(bank)} puzzles on the NORMAL net (p1 top-1 + p2 all-7)", flush=True)

    t0 = time.time()
    model = kr.load_model(os.path.join(kr.BT_MODELS, 'model-v1.0.0-normal.pth'))
    verdicts = []
    for i, pz in enumerate(bank):
        v = verdict_for(model, pz)
        verdicts.append(v)
        tag = 'KEEP' if v['keep'] else f"DROP/{v['reason']}"
        p2_tag = f" p2={v['p2_agree']}/{v['p2_of']}" if v['p2_of'] is not None else ''
        print(f"  {i+1}/{len(bank)} #{v['number']} {tag} rank={v['rank']}{p2_tag}", flush=True)

    json.dump(verdicts, open(out_path, 'w'))

    kept = sum(1 for v in verdicts if v['keep'])
    disagree = sum(1 for v in verdicts if v['reason'] == 'disagree')
    disagree_p2 = sum(1 for v in verdicts if v['reason'] == 'disagree-p2')
    unreachable = sum(1 for v in verdicts if v['reason'] == 'unreachable')
    bt_error = sum(1 for v in verdicts if v['reason'] == 'bt-error')
    other = len(verdicts) - kept - disagree - disagree_p2 - unreachable - bt_error
    n = len(verdicts)
    pct = lambda k: f"{k}/{n} ({100*k/n:.0f}%)" if n else f"{k}/0"
    print(f"\n==== CONSENSUS [normal, p1 top-1 + p2 all-7] n={n} ====")
    print(f"   KEEP (p1 top-1 + p2 {P2_THRESHOLD}/7):           {pct(kept)}")
    print(f"   DROP / disagree-p1 (reachable, rank>1):  {pct(disagree)}")
    print(f"   DROP / disagree-p2 (p1 ok, p2 <{P2_THRESHOLD}/7): {pct(disagree_p2)}")
    print(f"   DROP / unreachable:                      {pct(unreachable)}")
    print(f"   DROP / bt-error (counted separately):    {pct(bt_error)}")
    print(f"   DROP / other (odd-parity, inject):       {pct(other)}")
    # p2 agreement distribution for puzzles that passed p1
    p2_counts = [v.get('p2_agree') for v in verdicts if v.get('p2_of') is not None]
    if p2_counts:
        for t in range(8):
            c = sum(1 for x in p2_counts if x == t)
            if c:
                print(f"   p2 agree={t}/7: {c}")
    print(f"   wrote {out_path}; elapsed {int(time.time()-t0)}s", flush=True)


if __name__ == "__main__":
    main()
